Add 'pip cache' command

duckinator opened the pull request:
<!---
Thank you for your soon to be pull request. Before you submit this, please
double check to make sure that you've added a news file fragment. In pip we
generate our NEWS.rst from multiple news fragment files, and all pull requests
require either a news file fragment or a marker to indicate they don't require
one.

To read more about adding a news file fragment for your PR, please check out
our documentation at: https://pip.pypa.io/en/latest/development/contributing/#news-entries
-->

This PR adds `pip cache` command, as per #4685.

TODO:
- [x] Fix tests
- [x] General refactoring/cleanup
- [x] Documentation
- [x] Add news file fragment
- [x] Add test for `pip cache list` with an empty cache directory.
- [x] Add test for `pip cache list` that verifies the number of items in the cache is reported correctly.
- [x] Add test for `pip cache purge` which verifies that it (if called correctly) deletes everything in the wheel cache.
- [x] Add test for `pip cache purge` which verifies that it raises an error + doesn't delete anything if given too many arguments.
- [x] Convert repeated section to a fixture? See https://github.com/pypa/pip/pull/6391#discussion_r388114107.
- [ ] Open issue about adding an option to control sorting of `pip cache list` (mainly, allow sorting by size).

(Thanks to ~@BrownTruck~ @xavfernandez and @tdsmith for their attempts, in #3734 and #3968 respectively, which I took a lot of inspiration/ideas from.)

duckinator commented:
I'm going AFK before the tests finish, but _theoretically_ the only failing tests this time should be related to my use of `options.cache_dir` in `CacheCommand.run()`.

Looking forward to seeing how wrong I am about that. :rofl:

duckinator commented:
Apparently squashing my commits is causing `isort` (in the lint tasks on Travis CI) some distress. Whoops.

duckinator commented:
4 failed tests + the weirdness with the linter on Travis CI.

First failed test: https://ci.appveyor.com/project/pypa/pip/builds/23692211/job/mephfb9qc9ida156#L229

```
[gw1] win32 -- Python 2.7.15 c:\projects\pip\.tox\py\scripts\python.exe
script = <tests.lib.PipTestEnvironment object at 0x04A759F0>
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x04A757B0>
    def test_cache_info(script, monkeypatch):
        result = script.pip('cache', 'info')
    
        cache_dir = appdirs.user_cache_dir('pip')
    
>       assert 'Location: %s' % cache_dir in result.stdout
E       AssertionError: assert ('Location: %s' % 'C:\\Users\\appveyor\\AppData\\Local\\pip\\Cache') in 'Cache info:\n  Location: t:\pytest-0\popen-gw1\test_cache_info0\home\AppData\Local\pip\Cache\n  Packages: 0\n'
E        +  where 'Cache info:\n  Location: t:\pytest-0\popen-gw1\test_cache_info0\home\AppData\Local\pip\Cache\n  Packages: 0\n' = <tests.lib.TestPipResult object at 0x04A75110>.stdout
tests\functional\test_cache.py:12: AssertionError
```

As I mentioned above, there's a disagreement between `options.cache_dir` in `CacheCommand.run()` and `appdirs.user_cache_dir('pip')` in the test.

I think the value of `options.cache_dir` is what should be being used here, which means I need to find out where that value comes from.

duckinator commented:
An observation:

- `options.cache_dir` appears to -- unless `--cache-dir` is passed -- default to `pip._internal.locations.USER_CACHE_DIR`... which is defined as `appdirs.user_cache_dir('pip')`
- Calling `appdirs.user_cache_dir('pip')` in my test gives a _different result_.

From poking around, I _think_ `script.pip(...)` is creating a subprocess, and modifies the environment when doing so. This would also explain the discrepancy, but I'm still not sure how to actually resolve the problem.

(Aside: looks like commands should return a status, so I'll need to change my code to do that.)

pradyunsg commented:
Thanks for this PR @duckinator! :D

> I think `script.pip(...)` is creating a subprocess, and modifies the environment when doing so. 

Yep. I can confirm this.

---

> @BrownTruck

That PR was by @xavfernandez that @BrownTruck (a bot) rebased. No worries on getting this detail wrong tho. :P

duckinator commented:
FYI, I'm still happy to work on this, but I haven't been able to figure out how to get the correct cache directory. That's the main hitch with me wrapping this up.

If anyone has any ideas on how to get that, I'd greatly appreciate it.

pradyunsg commented:
Hey @duckinator!

This is still on my radar; I just haven't had much time outside of RL to get to answering your question.

duckinator commented:
Rebased, but not made any changes.

xavfernandez commented:
Tests are using https://github.com/pypa/pip/blob/master/tests/conftest.py#L98-L166 which should help you find the correct cache directory.

chrahunt commented:
The basic issue appears to be that changes to `APPDATA` and `LOCALAPPDATA` in the environment of the current process do not impact the returned value of `ctypes.windll.shell32.SHGetFolderPathW`. Only for sub-processes.

One option could be to take the current value of `os.environ['LOCALAPPDATA']` and append `\pip\Cache` to it, duplicating the logic from `appdirs.user_cache_dir`.

Another option could be to get the cache dir from a sub-process itself, like `script.run('python', '-c', 'from pip._internal.locations import USER_CACHE_DIR; print(USER_CACHE_DIR)`. This seems like more of a hack, but does avoid the duplication.

In either case, I may be good to make it a property of `PipTestEnvironment` so it's accessible from the required tests or any others that might need the cache directory.

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will be eligible for code review and hopefully merging!

duckinator commented:
I'll try to get around to rebasing this and updating it taking into account @chrahunt's comment soon. Sorry for the delay!

duckinator commented:
Rebased and fixed a minor formatting problem which was causing a linting error.

I'm currently waiting for CI to finish to see if there's any new/unexpected errors. Once that's all sorted, I'm going to try this approach:

> One option could be to take the current value of os.environ['LOCALAPPDATA'] and append \pip\Cache to it, duplicating the logic from appdirs.user_cache_dir.

As @chrahunt suggested, I think moving it to `PipTestEnvironment` would make sense, but I think I'll make a separate issue/PR (depending on how ambitious I feel :sweat_smile: ) instead. This PR is relatively large as it is.

duckinator commented:
I opened issue #7082 about accessing environment variables that are modified in subprocesses.

duckinator commented:
Okay, looks like I'm not going to be able to add a fix for the Windows tests tonight, but I rebased my changes, opened #7082, and have a rough idea of how I want to approach it.

pradyunsg commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,105 @@
+        purge:
+            Remove all items from the cache.
+    """
+    actions = ['info', 'list', 'remove', 'purge']
```
Hmm, since this is another command with sub-commands (`pip config` is the other one), would it be OK with you to adapt this to that model for sub-command handling?

I imagine there would be two ways to go about it -- copy paste the code here, keeping the core logic for dispatching-to-subcommands the same (which we can refactor in a follow up PR) or move out the core logic for dispatching-to-subcommands into a dedicated class and then follow up with this PR being updated to do that.

I'm fine with either (with a preference for the latter, because it's cleaner) but I can definitely understand if you just want this PR to be just-add-the-cache command (i.e. the former) and we can look into refactoring to reduce the technical debt later. :)

duckinator replied:
@pradyunsg I'm planning to poke at this again next week. I'll look into doing the second approach you mentioned when I do so. :+1:

xavfernandez commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,105 @@
+            pattern = '*'
+
+        files = self.find_wheels(pattern)
+        wheels = map(self._wheel_info, files)
```
I'm not convinced that hiding the python, abi & platform tags (via `self._wheel_info`) is a good idea...
One use case would be to quickly identify all Python 3.5 wheels once you know you won't work with it anymore

duckinator replied:
Looking at it again, I'm not convinced it is, either. iirc it's the only way I could get it to work nicely, but I don't recall the details of why that is. Will need to look into it more.

duckinator replied:
@xavfernandez my latest changes make it so it just shows the entire file name, but _wheel_info still removes the directory. e.g.

```
$ python3 src/pip cache list
Current cache contents:
 - PyYAML-5.1.2-cp27-cp27mu-linux_x86_64.whl
 - PyYAML-5.1.2-cp37-cp37m-linux_x86_64.whl
 - functools32-3.2.3.post2-cp27-none-any.whl
 - mypy_extensions-0.4.2-py2.py3-none-any.whl
 - nodeenv-1.3.3-cp37-none-any.whl
 - pip-19.3.dev0-cp27-none-any.whl
 - pip-19.3.dev0-cp37-none-any.whl
 - scandir-1.10.0-cp27-cp27mu-linux_x86_64.whl
 - scripttest-1.3-cp27-none-any.whl
 - scripttest-1.3-cp37-none-any.whl
 - snowballstemmer-1.9.1-cp37-none-any.whl
$
```

duckinator commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,105 @@
+
+        for filename in files:
+            os.unlink(filename)
+        logger.info(result.strip())
```
yeah, agreed. seems like i kinda overengineered that bit.

duckinator replied:
Switched to this. It now prints `Removed <number of files> files` by default, and if you've passed `--verbose` (or otherwise cause `logger.debug()` to print things) it should print `Removed <filename>` for each filename, as well.

duckinator commented:
!!! all the tests are passing on all platforms! finally!

all remaining CI failures are accounted for by lack of a news entry and linter errors due to a lack of type annotations.

i'm leaving the type annotations out until after addressing the changes @pradyunsg mentioned, since that'll likely involve a bunch of changes.

duckinator commented:
@pradyunsg I switched it over to the style that `pip config` uses for subcommands. It looks like the overlap is actually pretty small &mdash; everything that's duplicated is isolated to 12 lines of code in `run()`.

However, `ConfigurationCommand.run()` has a chunk of command-specific code in the middle of the common parts, so I think I'd rather set aside consolidating things until after this is merged, since it feels like it may be more complex than it looks on the surface.

pradyunsg commented:
That's perfectly fine with me. :)

Thanks for taking the time to make the change! I'll try to get to reviewing this by the weekend, after my talk at PyCon India.

duckinator commented:
Okay, tests passed, so this is ready for review!

pradyunsg commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,129 @@
+
+    def list_cache_items(self, options, args):
+        # type: (Values, List[Any]) -> None
+        if args and args[0]:
```
This should ensure the length of `args` is 0 or 1 on entry into this handler. Then, this conditional can be simplified to just be `if args:`.

pradyunsg commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,129 @@
+
+    def remove_cache_items(self, options, args):
+        # type: (Values, List[Any]) -> None
+        if not args:
```
This should ensure len(args) == 1

pradyunsg commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,129 @@
+            logger.debug('Removed %s', filename)
+        logger.info('Removed %s files', len(files))
+
+    def purge_cache(self, options, args):
```
This should ensure `not args`.

pradyunsg commented on tests/functional/test_cache.py:
```diff
@@ -0,0 +1,69 @@
+
+
+def _cache_dir(script):
+    results = script.run(
```
```suggestion
    result = script.run(
```

pradyunsg commented on tests/functional/test_cache.py:
```diff
@@ -0,0 +1,69 @@
+        'from pip._internal.locations import USER_CACHE_DIR;'
+        'print(USER_CACHE_DIR)'
+    )
+    return str(results.stdout).strip()
```
```suggestion
    return result.stdout.strip()
```

Why was this wrapped with `str`?

duckinator replied:
Unintentional leftover from my first attempt before I remembered that, uh, reading the docs is useful. :sweat_smile:

duckinator commented:
@pradyunsg super useful feedback, will make the appropriate changes a bit later today. :+1:

xavfernandez commented on docs/man/commands/cache.rst:
```diff
@@ -0,0 +1,20 @@
+:orphan:
```
The same kind of basic file would also be needed in https://github.com/pypa/pip/tree/master/docs/html/reference

xavfernandez commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,129 @@
+        results = textwrap.dedent("""\
+            Cache info:
+              Location: %s
+              Packages: %s""" % (options.cache_dir, num_packages)
```
Since we are only inspecting the `wheels` cache (and not the http part), I'd say this should appear in the output.

xavfernandez replied:
And like in #3734, as an user I would be interested to know the total size used.

duckinator replied:
> Since we are only inspecting the `wheels` cache (and not the http part), I'd say this should appear in the output.

@xavfernandez do you mean that it should say the location is e.g. (on my system) `/home/puppy/.cache/pip/wheels` instead of `/home/puppy/.cache/pip`?

> And like in #3734, as an user I would be interested to know the total size used.

Yup, this would be good! I'll look into it!

xavfernandez commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,129 @@
+    def _find_wheels(self, options, pattern):
+        # type: (Values, str) -> List[str]
+        wheel_dir = os.path.join(options.cache_dir, 'wheels')
+        return find_files(wheel_dir, pattern + '-*.whl')
```
Why the `-*.whl` postfix (vs a simple `.whl`) ? If I want to find all `*-linux_x86_64.whl` files ?

duckinator replied:
Good catch. I'll change this to `pattern + '*.whl'`, since just `pattern + '.whl'` would would mean that e.g. `pip list PyYAML` would return nothing, but `pip list PyYAML*` would work as it does now.

duckinator commented:
TODO:
- [ ] `textwrap.dedent`-related stuff
- [ ] verify number of args in `list_cache_items`
- [ ] verify number of args in `remove_cache_items`
- [ ] verify number of args in `purge_cache`
- [ ] rename `results`  to `result` and avoid needless `str()`
- [ ] add file to docs/html/reference
- [ ] https://github.com/pypa/pip/pull/6391#discussion_r332995278
- [ ] include total sized used in `pip cache info`
- [ ] change `pattern + '-*.whl'` to `pattern + '*.whl'`

duckinator commented:
@pradyunsg @xavfernandez I believe I handled all of your comments, aside from including size information.

duckinator commented:
I'll probably work on including cache size information this weekend or next week.

The rest of this is mostly a reminder for myself when I get back to working on it.

---

I'm planning to add the following:
- `pip cache info` include the total size of all wheel files in the cache
- `pip cache list` include the size of each file

For each of these, it'll round to the most appropriate unit out of B, KB, MB, or GB.

Jamim commented:
Hello @duckinator,

Thank you for this PR!

> For each of these, it'll round to the most appropriate unit out of B, KB, MB, or GB.

Probably, it's worth using [binary prefixes](https://en.wikipedia.org/wiki/Binary_prefix) such as `Ki`, `Mi`, `Gi` (e.g. `KiB`, `MiB`, `GiB`) for the sake of clarity. The point is, when you see `1 MB` you have to guess (or even figure out) if it's `10^6 bytes` or `2^20 bytes`. And when you see `1 MiB` you can be sure (more or less) it's `2^20`.

duckinator commented:
@Jamim sounds good to me. :+1:

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will be eligible for code review and hopefully merging!

hugovk commented:
This looks great!

---

Because pip uses a [different cache directory depending on OS](https://pip.pypa.io/en/stable/reference/pip_install/#caching), and because pip has no API, it means that caching with GitHub Actions is rather verbose.

It needs 3 conditional builds step, one for each `path: <PIP_CACHE_DIR_FOR_OS>`:

```yml
- uses: actions/cache@v1
  if: startsWith(runner.os, 'Linux')
  with:
    path: ~/.cache/pip
    key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements.txt') }}
    restore-keys: |
      ${{ runner.os }}-pip-

- uses: actions/cache@v1
  if: startsWith(runner.os, 'macOS')
  with:
    path: ~/Library/Caches/pip
    key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements.txt') }}
    restore-keys: |
      ${{ runner.os }}-pip-

- uses: actions/cache@v1
  if: startsWith(runner.os, 'Windows')
  with:
    path: ~\AppData\Local\pip\Cache
    key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements.txt') }}
    restore-keys: |
      ${{ runner.os }}-pip-
```

One idea is to use `python -c "from pip._internal.locations import USER_CACHE_DIR; print('::set-output name=dir::' + USER_CACHE_DIR)"` to get the cache directory (https://github.com/actions/cache/pull/14#issuecomment-550501012), but obviously as pip has no API it's not recommended to use pip's internal functions.

But a pip command like, for example, `pip cache dir` to return just the directory would be really useful!

What do you think? Would it be in scope for this PR?

I see `pip cache info` returns the wheels directory, plus some extra info, so it'd be like a cut down version of that. For example:

```console
$ pip cache info
Cache info:
  Location: /Users/hugo/Library/Caches/pip/wheels
  Packages: 471
$ pip cache dir
/Users/hugo/Library/Caches/pip
$ python -c "from pip._internal.locations import USER_CACHE_DIR; print(USER_CACHE_DIR)"
/Users/hugo/Library/Caches/pip
```

Thanks!

chrahunt commented:
Hi @hugovk! You can provide `--cache-dir` or set the `PIP_CACHE_DIR` environment variable to whatever path you like and it should be respected. If that's not enough I would raise an issue that could be addressed in a separate PR.

pradyunsg commented:
@duckinator Would you have the time to work on this anytime in the future? Is the current blocker that this PR needs a review from someone?

duckinator commented:
@pradyunsg hey! Sorry for disappearing there.

I believe all that's left is including size information (as explained in https://github.com/pypa/pip/pull/6391#issuecomment-540623035 and https://github.com/pypa/pip/pull/6391#issuecomment-540663877 ), and then having someone review it.

I'm busy tonight/tomorrow, but will try to wrap this up ASAP. I think it's pretty close.

hugovk commented:
Re: KiB vs. KB https://github.com/pypa/pip/pull/6391#issuecomment-540663877

Whilst KiB, MiB, GiB, etc., may be technically less ambiguous, I don't think it adds clarity and will confuse some people.

Hard disk manufacturers use things like 10^6, but operating systems are pretty uniform in using 1 MB = 2^20.

And it would look rather strange to suddenly see different units when doing `pip cache list` compared with `ls` and the rest of the OS.

uranusjr commented:
Why do we need to choose though? In `ls` (since it is mentioned) the user can choose to use either `M` and `MB`, and we can just do the same.

pradyunsg commented:
> I'm busy tonight/tomorrow, but will try to wrap this up ASAP. I think it's pretty close.

Good to know! I'll try to come around to reviewing this later this month (feel free to ping me if I don't).

BTW, if you think you wouldn't be able to make time for this in the near future, that's perfectly fine. Let us know if that's the case -- that way someone else (possibly me?) can pick this up and take it forward if needed.

pradyunsg commented:
Worth flagging that #7350 is a feature request for this PR. :)

pfmoore commented:
> Why do we need to choose though?

Well, I'd be very much against having an option for the user to choose:

1. We'd still have to pick a default, so we don't actually avoid the discussion on what's preferable.
2. One more choice to confuse the user.
3. Having such an option only affect this one command, and not work globally, would be very confusing (does pip use K/M prefixes anywhere else? I don't know TBH).

My preference is just to use K/M/G as @duckinator originally proposed. I doubt many people will be looking at the figure as anything other than an indicative value, anyway.

hugovk commented:
> (does pip use K/M prefixes anywhere else? I don't know TBH)

Searching the code I only found this, in a vendored library:

https://github.com/pypa/pip/blob/ccd2ced9bccedf5f3f0c1d23eb7bdd2580f1c8c7/src/pip/_vendor/distlib/util.py#L1270

hugovk commented:
Currently `git cache purge` removes the `*.whl` files from the `pip/wheels/` directory.

Should it remove the `pip/wheels/` directory itself so all the subdirs are also cleaned up? In my case, it's only 57 KB afterwards, but I guess I don't need those at all.

And I also still have a 2.1 GB `pip/http/` directory after purge. Should that also be removed?

And `selfcheck.json` and `pip/selfcheck/` are only 27 KB, but should `purge` remove the whole `pip/` cache dir?

I think my `pip/wheels/` directory was quite small, especially when compared with `pip/wheels/`.

chrahunt commented:
IMO we can tackle those optimizations as a separate issue - if we start deleting directories then we would want to consider possible race conditions with other pip processes.

duckinator commented:
> Good to know! I'll try to come around to reviewing this later this month (feel free to ping me if I don't).
> 
> BTW, if you think you wouldn't be able to make time for this in the near future, that's perfectly fine. > Let us know if that's the case -- that way someone else (possibly me?) can pick this up and take it > forward if needed.

@pradyunsg: Aside from the last few days, I usually have plenty of time. The problem is I am _really_ bad at remembering things lmao

I got my computer set back up tonight, so it's just a matter of me managing to remember to get back to it. :joy: 

> Worth flagging that #7350 is a feature request for this PR. :)

@pradyunsg: Yup! Left a comment on there. I like the idea, but think it'd work best as a follow-up PR for this one given how big this one already is.

---

> Whilst KiB, MiB, GiB, etc., may be technically less ambiguous, I don't think it adds clarity and will confuse some people.

@hugovk: Thinking about it more, I'm inclined to agree with you, especially given (as @pfmoore hinted at) I'd expect it to be used for getting a rough idea of the size of things as opposed to the exact number of bytes.

hugovk commented:
> IMO we can tackle those optimizations as a separate issue - if we start deleting directories then we would want to consider possible race conditions with other pip processes.

I've created #7372 for #6391 (comment).

hugovk commented:
@duckinator Happy new year! Let me know if I can help with this, it'd be great to include it in the 31 January release.

pradyunsg commented:
@hugovk Noting that I'm personally OK with not including this in the Jan 31 release, partly because this PR still needs updates IIRC and we'd still have to review a decently sized PR here.

Honestly though, there's no hurry here and we can easily defer this to the next release (that's part of the point of a fairly regular and frequent cadence).

duckinator commented:
Hi, sorry for the lack of response. Been dealing with some health issues recently. I think it's finally ready for review!

hugovk commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,152 @@
+            wheel = os.path.basename(filename)
+            size = filesystem.friendly_file_size(filename)
+            results.append(' - {} ({})'.format(wheel, size))
+        logger.info('Current cache contents:\n')
```
We only care about now, and more concise:
```suggestion
        logger.info('Cache contents:\n')
```

hugovk commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,152 @@
+        files = self._find_wheels(options, pattern)
+
+        if not files:
+            logger.info('Nothing is currently cached.')
```
We only care about now, and more consice:

```suggestion
            logger.info('Nothing is cached.')
```

hugovk commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,152 @@
+
+class CacheCommand(Command):
+    """
+        Inspect and manage pip's caches.
```
Here and some other places it's plural, others it's singular. Perhaps be consistent and stick with singular "cache"? Or is it referring to different types of cache?

duckinator replied:
I'm not entirely sure why I was so inconsistent about it, but I'm changing it to always be singular.

hugovk commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,152 @@
+        for filename in files:
+            os.unlink(filename)
+            logger.debug('Removed %s', filename)
+        logger.info('Removed %s files', len(files))
```
Perhaps also deal with singular case? So:
```
Removed 1 file
Removed 2 files
```
Instead of:
```
Removed 1 files
Removed 2 files
```

chrahunt replied:
Alternatively, "file(s)".

hugovk replied:
Computers are bad at many things, but one thing they're really good at is counting :)

uranusjr replied:
Personally I prefer file(s). I am probably super biased since my native language does not have plurals 😛  but IMO locale-sensitive string formatting is bad for logging in general.

duckinator replied:
Changing it to `file(s)`.

hugovk replied:
I'd like to suggest not to use "file(s)".

If it absolutely must be static for logging, may I suggest rewriting like this?

```diff
-        logger.info('Removed %s files', len(files))
+        logger.info('Files removed: %s', len(files))
```

Here are some style guides that say to avoid "(s)":

* https://developers.google.com/style/plurals-parentheses

![image](https://user-images.githubusercontent.com/1324225/72334880-b8cba600-36c6-11ea-8278-5262ab55c3ad.png)


* IBM Style Guide:
![image](https://user-images.githubusercontent.com/1324225/72334587-3642e680-36c6-11ea-82b5-318978196acb.png)

* Global English Style Guide: Writing Clear, Translatable Documentation for a Global Market:

![image](https://user-images.githubusercontent.com/1324225/72334625-48bd2000-36c6-11ea-98b5-33f5b03ec670.png)

uranusjr replied:
*Files removed* works for me.

uranusjr replied:
@hugovk That works for me.

duckinator replied:
@hugovk I appreciate your feedback here. I'm changing it to `Files removed: %s` like you suggested. :+1:

hugovk commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,152 @@
+
+        for filename in files:
+            os.unlink(filename)
+            logger.debug('Removed %s', filename)
```
Use `.format`?
```suggestion
            logger.debug('Removed {}'.format(filename))
```

chrahunt replied:
If not logged then it doesn't incur the penalty that `.format`would.

duckinator replied:
I'm unsure if the reason @chrahunt mentioned is why I did this originally, but I'm in favor of leaving it for that reason. :sweat_smile:

hugovk commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,152 @@
+        for filename in files:
+            os.unlink(filename)
+            logger.debug('Removed %s', filename)
+        logger.info('Removed %s files', len(files))
```
Use `.format`?
```suggestion
        logger.info('Removed {} files',format(len(files)))
```

chrahunt replied:
Same as above, this avoids overhead if needed t logged.

hugovk commented on tests/functional/test_cache.py:
```diff
@@ -0,0 +1,101 @@
+    result = script.pip('cache', 'info')
+    cache_dir = _cache_dir(script)
+
+    assert 'Location: %s' % cache_dir in result.stdout
```
Use `.format`?
```suggestion
    assert 'Location: {}'.format(cache_dir) in result.stdout
```

hugovk commented on docs/html/reference/pip_cache.rst:
```diff
@@ -0,0 +1,22 @@
+.. _`pip cache`:
+
+pip cache
+------------
```
```suggestion
---------
```

hugovk commented on docs/man/commands/cache.rst:
```diff
@@ -0,0 +1,20 @@
+:orphan:
+
+==========
```
```suggestion
=========
```

hugovk commented on docs/man/commands/cache.rst:
```diff
@@ -0,0 +1,20 @@
+
+==========
+pip-cache
+==========
```
```suggestion
=========
```

duckinator commented:
@hugovk I think I addressed everything you mentioned! :slightly_smiling_face:

duckinator commented:
@hugovk my last commit changed `Removed %s file(s)` to `Files removed: %s` like you suggested earlier today. :+1:

duckinator commented:
Rebased off master.

duckinator commented:
I'm not sure why rebasing it caused all the Windows tests to start failing.

---

It looks like this is the problem for some of them:

```
AssertionError: assert 'Location: r:\\temp\\pytest-of-VssAdministrator\\pytest-1\\popen-gw0\\test_cache_info0\\home\\AppData\\Local\\pip\\Cache' in 'Cache info:\n  Location: r:\temp\pytest-of-vssadministrator\pytest-1\popen-gw0\test_cache_info0\home\appdata\local\pip\cache\wheels\n  Packages: 0\n  Size:     0.0 B\n'
```

And then this is the problem for some others:

```
AssertionError: assert 'Location: R:\\Temp\\pytest-of-VssAdministrator\\pytest-1\\popen-gw1\\test_cache_info0\\home\\AppData\\Local\\pip\\Cache' in 'Cache info:\n  Location: r:\\temp\\pytest-of-vssadministrator\\pytest-1\\popen-gw1\\test_cache_info0\\home\\appdata\\local\\pip\\cache\\wheels\n  Packages: 0\n  Size:     0.0 B\n'
```

It seems like some kind of strange escaping problem _and_ a capitalization problem?

duckinator commented:
It looks like that's from this line:

```
    assert 'Location: {}'.format(cache_dir) in result.stdout
```

So I think the problem is that `cache_location = self._wheels_cache_dir(options)` and `pip._internal.locations.USER_CACHE_DIR` are normalized differently.

uranusjr commented:
Try `assert 'Location: {}'.format(os.path.normcase(cache_dir)) in result.stdout`

sbidoul commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,152 @@
+
+class CacheCommand(Command):
+    """
+        Inspect and manage pip's cache.
```
```suggestion
        Inspect and manage pip's wheel cache.
```

To hint at the existence of the http cache, which this command does not inspect?

duckinator replied:
Just made a commit updating all 3 sentences I'd added that used the phrase “pip's cache” to say “pip's wheel cache”. :+1:

sbidoul commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,152 @@
+    usage = """
+        %prog info
+        %prog list [name]
+        %prog remove <pattern>
```
Does the list command accept a pattern too?

duckinator replied:
Yeah, it does. How would that be represented, since it's optional? `[<pattern>]`, `[pattern]`, or what?

sbidoul replied:
Looking at pip's style for these things I'd say `[<pattern>]` but I've no strong opinion on this.

duckinator replied:
Switched it to `%prog list [<pattern>]`. :+1:

duckinator commented:
Thanks @uranusjr. Looks like that fixed it.

uranusjr commented on src/pip/_internal/utils/filesystem.py:
```diff
@@ -169,3 +170,62 @@ def _test_writable_dir_win(path):
+
+    size = round(size, 1)
+
+    return '{} {}'.format(size, suffix)
```
This might return a misleading result on Python 2 when `size` is an int (yeah, Python 2…)

duckinator commented:
Thanks for the feedback, @sbidoul and @uranusjr! The problems you mentioned should be resolved now. :slightly_smiling_face:

duckinator commented:
Forgot to mention &mdash; I also rebased off master when working on this earlier today.

pradyunsg commented:
Alright. Reviewing this now.

pradyunsg commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,152 @@
+    def _find_wheels(self, options, pattern):
+        # type: (Values, str) -> List[str]
+        wheel_dir = self._wheels_cache_dir(options)
+        return filesystem.find_files(wheel_dir, pattern + '*.whl')
```
Maybe, we should allow the users to skip the "*" here -- which might be relevant if they want an exact match?

pfmoore replied:
For `list`, this is a relatively minor point, and I wouldn't be inclined to worry much. But in the case of `remove`, is prefix search what we want at all? Consider:

```
>.\.venv\Scripts\pip cache list
Cache contents:

 - MarkupSafe-1.1.1-cp38-cp38-win_amd64.whl (19 kB)
 - coverage-5.0.3-cp38-cp38-win_amd64.whl (205 kB)
 - distlib-0.3.0-cp38-none-any.whl (340 kB)
 - env_var_test-0.1-cp38-none-any.whl (1.3 kB)
 - env_var_test-0.1-cp38-none-any.whl (1.3 kB)
 - flit-2.2.0-cp38-none-any.whl (49 kB)
 - flit_core-2.2.0-cp38-none-any.whl (40 kB)
 - foo-0.1-cp38-none-any.whl (1.2 kB)
 - foo-0.1-cp38-none-any.whl (1.2 kB)
 - foo-0.1-cp38-none-any.whl (1.2 kB)
 - importlib_metadata-1.4.0-cp38-none-any.whl (29 kB)
 - msgpack-0.6.2-cp38-cp38-win_amd64.whl (14 kB)
 - pip-20.0.dev0-cp38-none-any.whl (1.4 MB)
 - pip-20.1.dev0-py2.py3-none-any.whl (1.4 MB)
 - pip-20.1.dev0-py2.py3-none-any.whl (1.5 MB)
 - pkg123-0.7-cp38-none-any.whl (1.2 kB)
 - pkg123-0.7-cp38-none-any.whl (1.2 kB)
 - pkg123-0.7-cp38-none-any.whl (1.2 kB)
 - pkg123-0.7-cp38-none-any.whl (1.2 kB)
 - pkg123-0.7-cp38-none-any.whl (1.2 kB)
 - pkg512-0.0.5-cp38-none-any.whl (1.2 kB)
 - pkgX-0.0.5-cp38-none-any.whl (1.2 kB)
 - pkgX-0.0.5-cp38-none-any.whl (1.2 kB)
 - pkgX-0.0.5-cp38-none-any.whl (1.2 kB)
 - progress-1.5-cp38-none-any.whl (8.1 kB)
 - proj-0.1-cp38-none-any.whl (1.2 kB)
 - pycparser-2.19-py2.py3-none-any.whl (111 kB)
 - pyrsistent-0.15.7-cp38-cp38-win_amd64.whl (71 kB)
 - retrying-1.3.3-cp38-none-any.whl (11 kB)
 - scripttest-1.3-cp38-none-any.whl (7.8 kB)
 - setuptools_scm-3.4.3-cp38-none-any.whl (26 kB)
 - tox-3.14.4.dev4+g581cc25.d20200116-cp38-none-any.whl (82 kB)
 - zipp-2.0.1-cp38-none-any.whl (4.2 kB)
PS 15:02 [Git: pr_6391] {00:00.289} C:\Work\Projects\pip
>.\.venv\Scripts\pip cache remove pkg
Files removed: 9
```

I'm pretty sure my instinct would have been that I meant to remove wheels for project `pkg`, not for `pkg123`, `pkg512`, or `pkgX`.

My inclination would be to make the argument a wildcard match on *just* the project name part of the wheel - specifically because of `remove`.

duckinator replied:
I agree that for `pip cache remove`, this is definitely a problem. I'm not sure how to handle it, though &mdash; the main point of the `*` is to allow you to do (using one from your results as an example) e.g. `pip cache remove tox` or `pip cache remove tox-3.14.4` instead of `pip cache remove tox-3.14.4.dev4+g581cc25.d20200116-cp38-none-any.whl`.

I guess `-*.whl` would be an improvement, at least. But can the name include a hyphen in `.whl` files? E.g., if a package is called `foo-bar`, is the wheel `foo-bar-<version>-<...>.whl`? If so, this still leaves the possibility of incorrect matches (like `pip remove foo` removing both `foo` and `foo-bar`).

duckinator replied:
Looking into this more, according to [PEP 427 -- The Wheel Binary Package Format 1.0](https://www.python.org/dev/peps/pep-0427/#file-name-convention):

> The wheel filename is `{distribution}-{version}(-{build tag})?-{python tag}-{abi tag}-{platform tag}.whl`.

This means, after `{distribution}`, there's either 4 or 5 hyphens depending on if there's a build tag. I think I may need to use two regular expressions (one with the build tag, one without) and just check if each filename matches _either_ of them, and include the ones that do?

pfmoore replied:
The details of wheel names are [in the PEP](https://www.python.org/dev/peps/pep-0427/#file-name-convention), but no, the project name can never contain a `-` (non-alphanumerics are [canonicalised](https://www.python.org/dev/peps/pep-0427/#escaping-and-unicode) to `_`) so the only place you have `-` is separating the parts.

We could do something like `pattern + ("*.whl" if "-" in pattern else "-*.whl")` - this would mean that a pure project name is matched exactly, but if the user specifies `name-version`, the version is treated as a prefix.

This seems like it's sufficient - I don't think we need to make this over-complicated at this stage. If we hit issues, they can easily enough be fixed in a follow-up PR that tightens things up based on real use cases. That's better than trying to either guess what could go wrong, or over-engineer things from the start.

(Even the `pip cache remove` case isn't *that* serious, if we delete more than the user intended, it's only a cache for performance, and the wheels will get rebuilt anyway, so there's no major damage caused).

duckinator replied:
@pfmoore I went ahead with the approach you suggested, and added a comment explaining the reasoning. And I agree, it's not a serious problem, but it's always nice for it to behave intuitively. Especially if the fix is this simple. :slightly_smiling_face:

pradyunsg commented on src/pip/_internal/utils/filesystem.py:
```diff
@@ -169,3 +170,67 @@ def _test_writable_dir_win(path):
+    # type: (str) -> Union[int, float]
+    # If it's a symlink, return 0.
+    if os.path.islink(path):
+        return 0
```
Arguably... symlinks do have a size -- the number of characters in the reference. :P

duckinator replied:
I was going off what `du` reports on my (Linux) system, but apparently different tools &mdash; and even different invocations of the same tool &mdash; report different things on my system!

E.g., in this case I just tried, `du` reports 0, `ls -s` reports 0, but `ls -l` reports 42 (the length of the reference):

```
~$ file ~/.config/nvim/init.vim
/home/puppy/.config/nvim/init.vim: symbolic link to /home/puppy/dotfiles/.config/nvim/init.vim
~$ du ~/.config/nvim/init.vim
0       /home/puppy/.config/nvim/init.vim
~$ ls -s ~/.config/nvim/init.vim
0 /home/puppy/.config/nvim/init.vim
~$ ls -l ~/.config/nvim/init.vim
lrwxrwxrwx. 1 puppy puppy 42 Jan 21 08:20 /home/puppy/.config/nvim/init.vim -> /home/puppy/dotfiles/.config/nvim/init.vim
~$
```

I _think_ what's happening is that, excluding `ls -l`, all the things I tried were rounding to the nearest KB.

I tried it on FreeBSD and got completely different results. On there, `du` reports 1(?!), `ls -s`  reports the length of the reference, and `ls -l` also reports the length of the reference.

So I'm not really sure what's most appropriate here. The two operating systems I have easy access to return different results, `os.path.getsize` resolves it (meaning it may count things multiple times), and I don't know what other operating systems do. :upside_down_face:

pfmoore replied:
I think it's probably irrelevant - this is only used in `pip cache` at the moment, which (as far as I know) won't contain symlinks anyway.

I'd suggest leaving it as it is. If someone comes along with a use case for this function that needs a different answer for symlinks, let's revisit it then.

duckinator replied:
Sounds good to me. Marking this thread as resolved. :+1:

pradyunsg commented on tests/functional/test_cache.py:
```diff
@@ -0,0 +1,101 @@
+    wheel_cache_dir = os.path.join(cache_dir, 'wheels')
+    destination = os.path.join(wheel_cache_dir, 'arbitrary', 'pathname')
+    os.makedirs(destination)
+    with open(os.path.join(destination, 'yyy-1.2.3.whl'), 'w'):
```
```suggestion
    with open(os.path.join(destination, 'yyy-1.2.3-py27-none-any.whl'), 'w'):
```

pfmoore replied:
nit: `-py3-none-any` - let's avoid Python 2 🙂

duckinator replied:
I went with one -py3 and one -py27 since that means if it works, it works on both versions, which is good to know. (Since pip still supports both.)

pradyunsg commented on tests/functional/test_cache.py:
```diff
@@ -0,0 +1,101 @@
+    os.makedirs(destination)
+    with open(os.path.join(destination, 'yyy-1.2.3.whl'), 'w'):
+        pass
+    with open(os.path.join(destination, 'zzz-4.5.6.whl'), 'w'):
```
```suggestion
    with open(os.path.join(destination, 'zzz-4.5.6-py27-none-any.whl'), 'w'):
```

pfmoore replied:
same nit - @pradyunsg I'm disappointed in you 😛

pradyunsg commented on tests/functional/test_cache.py:
```diff
@@ -0,0 +1,101 @@
+    with open(os.path.join(wheel_cache_dir, 'yyy-1.2.3.whl'), 'w'):
+        pass
+    with open(os.path.join(wheel_cache_dir, 'zzz-4.5.6.whl'), 'w'):
+        pass
```
Make this a fixture?

duckinator replied:
Agreed, but I'm not sure how fixtures work in the test setup pip uses, so focusing on the things I can get done quickly first. :+1:

pfmoore replied:
+1 on not worrying about this now.

duckinator replied:
I've not turned it into a fixture yet, but I have extracted it out into a function. So it's at least a lot less repetitive!

duckinator replied:
Converted it into a fixture. It may need more work, but it's definitely improved things significantly. :+1:

pradyunsg commented on tests/functional/test_cache.py:
```diff
@@ -0,0 +1,101 @@
+    with open(os.path.join(destination, 'yyy-1.2.3.whl'), 'w'):
+        pass
+    with open(os.path.join(destination, 'zzz-4.5.6.whl'), 'w'):
+        pass
```
Make this a fixture?

pradyunsg commented on tests/functional/test_cache.py:
```diff
@@ -0,0 +1,101 @@
+    assert 'yyy-1.2.3' not in result.stdout
+    assert 'zzz-4.5.6' not in result.stdout
+
+    result = script.pip('cache', 'purge', '--verbose')
```
I'd split these two calls out into separate tests; to isolate the checks from each other.

duckinator replied:
As a note to myself, I think I want to split this into two tests which test that:

- `cache purge <something with no matches> --verbose` raises an error _and_ doesn't delete anything. (the second part is tested implicitly by the second call working, but it'd be nice to _explicitly_ test it.)
- `cache purge --verbose` does not raise an error, deletes everything. (at the moment it only tests the output, but really it should check both that and that the file exists.)

and I'd like to add another test:

- `cache purge <something with matches> --verbose` does not raise an error, deletes all matching things, and _doesn't_ delete anything else.

pfmoore replied:
Those sound like very good tests 🙂

pradyunsg commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,152 @@
+            Cache info:
+              Location: {location}
+              Packages: {package_count}
+              Size:     {size}
```
```suggestion
            Location: {location}
            Size: {size}
            Number of wheels: {package_count}
```

duckinator replied:
Which of these do you think is better or more in-line with how pip typically does things?:

```
            Location: {location}
            Size: {size}
            Number of wheels: {package_count}
```

or:

```
            Location:         {location}
            Size:             {size}
            Number of wheels: {package_count}
```

pfmoore replied:
The first matches the behaviour of `pip show PKG`, and with the suggested change from "Packages:" to "Number of wheels:" I think the amount of whitespace is a little jarring. So I'd prefer the non-aligned version. I'm equally OK with your original version, though, TBH (in that case, aligning is OK because the lengths of the various headings are similar).

Ultimately, I think either is fine - this is for human consumption, not machine readability, so the key is readability which is always subjective.

duckinator replied:
I think I'll go with what @pradyunsg suggested because I also don't have a particularly strong opinion on it and it's consistent with other commands. There's discussion in another issue about adding a `--json` flag for getting a machine-readable version of `pip cache`, too, which would give us the freedom to tweak the human-readable format freely later on.

duckinator commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,152 @@
+        remove:
+            Remove one or more package from the cache.
+        purge:
+            Remove all items from the cache.
```
Good catch! The reason I didn't do this was because the original text I had was long enough to need wrapping. Didn't realize it was <79 characters after rephrasing things. :slightly_smiling_face:

duckinator commented on src/pip/_internal/commands/cache.py:
```diff
@@ -0,0 +1,152 @@
+            size = filesystem.friendly_file_size(filename)
+            results.append(' - {} ({})'.format(wheel, size))
+        logger.info('Cache contents:\n')
+        logger.info('\n'.join(sorted(results)))
```
Re sorting on (wheel, size): Yeah, it seems unlikely enough to not be worthwhile. We can always change that in the future if it turns out we're both wrong. :stuck_out_tongue: 

Re sort by size: Agree on all counts &mdash; seems useful, but I see no issue waiting until a follow-up PR for that. :slightly_smiling_face:

duckinator commented on src/pip/_internal/utils/filesystem.py:
```diff
@@ -169,3 +170,67 @@ def _test_writable_dir_win(path):
+    return result
+
+
+def _friendly_size(size):
```
Ah! I suspected there was something similar somewhere, but couldn't find it when I looked. Using that, and also renamed `friendly_{file,directory}_size` to `format_{file,directory}_size` for consistency. :+1:

duckinator commented:
I'm not sure what's causing the current CI failures. It says:

```
Executable `/Users/runner/.cache/pre-commit/repo14qw_y0i/py_env-python3.8/bin/python` not found
```

pfmoore commented:
> I'm not sure what's causing the current CI failures

I've seen that on another PR - I suspect it's an environment issue. Rerunning seems to have fixed it on the other PR, so I'll restart the tests here and keep my fingers crossed 🙂

duckinator commented:
TODO:

- [ ] Add test that specifies `--cache-dir` with an empty directory, and verifies it works.
- [ ] Add test that specifies `--cache-dir`, and verifies the number of files is actually correct.
- [ ] Add tests, as discussed in https://github.com/pypa/pip/pull/6391#discussion_r388312637.
- [ ] Convert repeated section to a fixture? See https://github.com/pypa/pip/pull/6391#discussion_r388114107.
- [ ] Open issue about adding an option to control sorting of `pip cache list` (mainly, allow sorting by size).

pradyunsg commented:
Ok great, looks like our pre-commit cache might be borked. Requesting @asottile for his expertize to chip in w.r.t. the pre-commit cache issues.

asottile commented:
hummm I'm not seeing cache related failures other than the lint-failing file contains the word cache:

```
Flake8...................................................................Failed
- hook id: flake8
- exit code: 1

tests/functional/test_cache.py:65:80: E501 line too long (80 > 79 characters)
tests/functional/test_cache.py:67:80: E501 line too long (81 > 79 characters)
tests/functional/test_cache.py:87:80: E501 line too long (80 > 79 characters)
tests/functional/test_cache.py:89:80: E501 line too long (81 > 79 characters)
```

asottile commented:
took one of the lines just to double check:

```
>>> len("     with open(os.path.join(wheel_cache_dir, 'yyy-1.2.3-py3-none-any.whl'), 'w'):")
81
```

pradyunsg commented:
Whoops! I'd meant to @ you in https://github.com/pypa/pip/pull/7824#event-3104820455; but looks like that was transient.

asottile commented:
ah, do you happen to have the failure handy somewhere? it looks like it passed and got merged :)

pradyunsg commented:
https://github.com/pypa/pip/runs/490837541

asottile commented:
huh interesting!  that looks like a combination of two things:
- github actions didn't cleanly restore the cache (looks like all the symlinks are broken in the cache venvs)
- and the [`healthy()`](https://github.com/pre-commit/pre-commit/blob/0a8ba31b9b6656d90f94fc368b47acb502cea44d/pre_commit/languages/python.py#L160-L168) check failed to consider this case -- looks like the healthy check should use a full path to the `python` executable to make sure to catch this type of corruption 🤔

hugovk commented:
I saw the same thing for the first in another repo this week too, worked around it by making a new cache: https://github.com/pylast/pylast/pull/322.

asottile commented:
I suspect actually that GA just moved where the tool cache is

On Fri, Mar 6, 2020, 12:57 PM Hugo van Kemenade <notifications@github.com>
wrote:

> I saw the same thing for the first in another repo this week too, worked
> around it by making a new cache: pylast/pylast#322
> <https://github.com/pylast/pylast/pull/322>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pypa/pip/pull/6391?email_source=notifications&email_token=AAN2BH35PRCKBUFXEMB3PITRGFPUVA5CNFSM4HELNFOKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEOC2T5I#issuecomment-595962357>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAN2BH4URVVZJJ7OBFZZ423RGFPUVANCNFSM4HELNFOA>
> .
>

asottile commented:
oops sorry didn't mean to hijack your PR -- we can continue the pre-commit discussion here: https://github.com/pre-commit/pre-commit/issues/1353

duckinator commented:
Not sure yet what the cause is, but it looks like on Windows, these tests are failing:

- `test_cache_remove_name_match`
- `test_cache_remove_name_and_version_match`
- `test_cache_purge`

They are, specifically, failing on the first non-negated `assert remove_matches_wheel(...)` call in each function. All of the `assert not remove_matches_wheel(...)` calls are fine. This combination makes me suspect that none of the lines are of the expected format, for some reason.

pradyunsg commented:
/cc @pfmoore and @uranusjr for their Windows expertize.

pfmoore commented on tests/functional/test_cache.py:
```diff
@@ -0,0 +1,217 @@
+
+        # The "/arbitrary/pathname/" bit is an implementation detail of how
+        # the `populate_wheel_cache` fixture is implemented.
+        expected = 'Removed {}/arbitrary/pathname/{}'.format(
```
The `/` characters should probably be `os.sep` - on Windows, `\` is the path separator.

pradyunsg replied:
Okay, now I feel stupid for @'ing you before taking a deeper look myself at the error.

pfmoore replied:
np, easy to miss :-) And I like @duckinator's fix of using `os.path.join` better than my suggestion anyway

duckinator replied:
Looks like that was it! Switched to `os.path.join()` instead of hard-coding the separator, which I suspect is functionally-equivalent, and it worked fine.

duckinator commented:
@pradyunsg @pfmoore I think this is ready for a re-review. :+1:

pradyunsg commented:
Close-reopened to retrigger CI. @duckinator many thanks for your patience here! We're on the final stretch now!

duckinator commented:
Rebased and resolved the merge conflict. /cc @pradyunsg

sbidoul commented:
@pradyunsg it would be great to have this one 20.1 too.

pradyunsg commented:
We're going to have a beta period for pip 20.1, and we can collect reports about how this works during the beta. Merging since I think it's very reasonable to have users try out this fairly new and big user facing chunk to our beta period and give us feedback. :)

Thanks for working on this @duckinator! This command has been a long time coming -- your patience and work on this is much appreciated! ^>^

pradyunsg commented:
Alsssooo, thanks everyone who reviewed this or made suggestions here! Much appreciated! ^>^

duckinator commented:
@pradyunsg I'm excited that it's finally been merged! Do you know what the timing for releasing the 20.1 beta will be?

pradyunsg commented:
See #7951 - it's April 20th. :)
